specs/data_structures.md

```diff
@@ -320,31 +286,178 @@ A proof into an SMT is structured as:
+| `messageSize`           | `uint64`                  | Size of message this transaction pays a fee for, in `byte`s. If this transaction does not pay a fee for a message, must be `0`.                                                                                                                                       |
+| `messageShareRootsRoot` | [HashDigest](#hashdigest) | Merkle root of message share roots of an optional message that this transaction pays a fee to be included in the current block. Messages are split into shares and committed to here. Large messages can span across rows, which requires more roots to the provided. |
+
+
+## IntermediateStateRootData
+
+| name                            | type                                                              | description                               |
+| ------------------------------- | ----------------------------------------------------------------- | ----------------------------------------- |
+| `wrappedIntermediateStateRoots` | [WrappedIntermediateStateRoot](#wrappedintermediatestateroot)`[]` | List of wrapped intermediate state roots. |
+
+### WrappedIntermediateStateRoot
+
+| name                    | type         | description                                                                                                     |
+| ----------------------- | ------------ | --------------------------------------------------------------------------------------------------------------- |
+| `index`                 | `uint64`     | Index of this intermediate state root in the list of intermediate state roots. This is needed for fraud proofs. |
+| `intermediateStateRoot` | `HashDigest` | Intermediate state root. Used for fraud proofs.                                                                 |
+
+### IntermediateStateRoot
+
+| name   | type         | description                                                                              |
+| ------ | ------------ | ---------------------------------------------------------------------------------------- |
+| `root` | `HashDigest` | Root of intermediate state, which is composed of the global state and the validator set. |
+
+## EvidenceData
+
+Wrapper for evidence data.
+
+| name        | type                      | description                                    |
+| ----------- | ------------------------- | ---------------------------------------------- |
+| `evidences` | [Evidence](#evidence)`[]` | List of evidence used for slashing conditions. |
+
+### Evidence
+
+| name     | type                    | description |
+| -------- | ----------------------- | ----------- |
+| `pubKey` | [PublicKey](#publickey) |             |
+| `voteA`  | [Vote](#vote)           |             |
+| `voteB`  | [Vote](#vote)           |             |
```

Comment:
If this is the only kind of evidence we will ever use (currently not 100% clear to me), we can be more specific in the naming everywhere (e.g. `DoubleSigningEvidence`)

Reply:
Thinking about this further, Evidence should probably be an enumerated type (currently only containing DoubleSigningEvidence as a member). Tendermint currently uses amino prefixing to encode the concrete instantiations of the Evidence interface but I think this will change to proto's `oneof` very soon.

Reply:
The particularities of evidence are outside the scope of this PR, but yes I agree that we'll need some evidence types (double signing is one, but we should also have one for proposing an invalid block---hence, fraud proofs).